### ISSUE TITLE
Option to enable RaycastAll for the Ray Perception Sensor Component

### ISSUE THREAD

**Issue by Xiromtz** [Unity-Technologies/ml-agents#3457]
In the game I'm using mlagents for I am currently testing with raycasts. I have a 2D top-down game where I raycast in the direction the agent is moving. Since it is a rather complicated game, there are lots and lots of colliders to check for. We have simple obstacle colliders the agent should check for, but at the same time there are more invisible trigger colliders like sightcones of enemies or traps only enemies can trigger. I want to be able to check for these trigger colliders, but at the same time be able to see whats behind them, e.g. some kind of obstacle the agent cannot pass through.
For example, when the agent is inside of an enemy sightcone, it won't be able to perceive ANYTHING outside of the sightcone itself. This is obviously very problematic.

There should be a simple option of using the RaycastAll feature to detect multiple tags with a single raycast. Though I don't know any implementation specifics, this might be simple to implement.

An obvious alternative is to use multiple perception components that interact with different layers, though this could get messy very quick.

Thanks!

**Comment by harperj** [comment]
Hi @Xiromtz -- thanks for the suggestion.  I agree that the best alternative at this time is to use multiple ray perception components.  I'll share your feedback with the team so that we can consider how to improve the ray perception feature.

**Comment by chriselion** [comment]
I'm not too familiar with `RaycastAll()`, but it looks like it can return an arbitrary number of results (depending on how long the ray is and what's in the way). On the other hand, observations need to be a fixed size. How do you propose deciding which results from RaycastAll get used in the observation? Use the closest N hits per ray, where N is a parameter on the sensor?

**Comment by surfnerd** [comment]
Hi @Xiromtz,
I'm curious about your comment 
```
For example, when the agent is inside of an enemy sightcone, it won't be able to perceive ANYTHING outside of the sightcone itself. This is obviously very problematic.
```

To me, it seems like your agent's raycasts should be ignoring your enemy's sight cone.  This can be set up in the layer mask of the agent's ray cast sensor. 

If you'd like you know the first N things a ray cast hit, then that is something we would need to work on.  Could you clarify what you are trying to achieve?

**Comment by Xiromtz** [comment]
Yeah at the time of writing I came up with the idea of using multiple raycast sensors and thought raycasting for multiple tags would be the optimal solution. 
The fixed size observations didn't occur to me since I don't know the raycast implementation. I guess every Ray has N - number of tags - observations, where each observation depicts the normalized distance to the tag found?
Using N number of possible hits per ray definetely sounds like a solution. After implementing multiple raycast sensors, it works really well and seems like the more sensible and flexible solution though. With the sensor detection on child objects feature, it's pretty easy to have multiple ray sensor components more clearly defined.

@surfnerd regarding youur question: My agent should be able to detect enemy sightcones á la stealth gameplay, but when already detected and wihin a sightcone, should still be able to detect objects in front of him. Hooking up the sightcones with its own layer and using a single ray perception sensor for that single layer works fine.

I am a bit worried about the overhead and in some usecases, we might want to detect objects behind others within the same layer. For these cases, the current solutions won't work.

In short: For me, using multiple ray cast sensors works perfectly fine, I might have prematurely written this request before thinking it more into detail. I do believe in some cases, having N collision per ray may be useful, albeit niche.

Thanks everyone for the time taken looking into this!
